hazelcast/src/main/java/com/hazelcast/sql/impl/physical/PhysicalNodeSchema.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.physical;
+
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Schema of a node.
+ */
+public class PhysicalNodeSchema implements FieldTypeProvider {
+
+    private final List<QueryDataType> types;
+    private final int rowWidth;
+
+    public PhysicalNodeSchema(List<QueryDataType> types) {
+        assert types != null;
+
+        this.types = Collections.unmodifiableList(types);
+
+        rowWidth = calculateRowWidth(types);
+    }
+
+    public static PhysicalNodeSchema combine(PhysicalNodeSchema schema1, PhysicalNodeSchema schema2) {
+        ArrayList<QueryDataType> types = new ArrayList<>(schema1.types);
+
+        types.addAll(schema2.types);
+
+        return new PhysicalNodeSchema(types);
+    }
+
+    @Override
+    public QueryDataType getType(int index) {
+        assert index <= types.size();
+
+        return types.get(index);
+    }
+
+    public List<QueryDataType> getTypes() {
+        return types;
+    }
+
+    public int getRowWidth() {
```

Comment:
Does it return an exact value, overestimates or underestimates? One may think it just returns the number of columns, probably a better name is something like `get(Over/Under)EstimatedSize`.

Reply:
This value is used to estimate memory requirements. We assume that it is relatively accurate. Bug again, this is not a concern of this ticket. The row width is currently used in IO handling, which will be prepared in a separate PR. I'll add JavaDocs with clarifications and possibly rename the method there. Otherwise, we will have unnecessary conflicts.